exchanges/binance/binance_websocket.go

```diff
@@ -601,55 +604,108 @@ func channelName(s *subscription.Subscription) (string, error) {
+
+	cNames := make([]string, len(chans))
+	for i := range chans {
+		c := chans[i]
+		cNames[i] = c.Channel
+		c.State = subscription.SubscribingState
+		if err := b.Websocket.AddSubscription(&c); err != nil {
+			return fmt.Errorf("%w Channel: %s Pair: %s Error: %w", stream.ErrSubscriptionFailure, c.Channel, c.Pair, err)
 		}
 	}
-	if len(payload.Params) > 0 {
-		err := b.Websocket.Conn.SendJSONMessage(payload)
-		if err != nil {
-			return err
+
+	req := WsPayload{
+		Method: wsSubscribeMethod,
+		Params: cNames,
+		ID:     id,
+	}
+
+	respRaw, err := b.Websocket.Conn.SendMessageReturnResponse(id, req)
+	if err == nil {
+		if _, d, _, rErr := jsonparser.Get(respRaw, "result"); rErr != nil {
+			err = rErr
+		} else if d != jsonparser.Null { // null is the only expected and acceptable response
+			err = errUnknownError
 		}
 	}
-	b.Websocket.AddSuccessfulSubscriptions(channelsToSubscribe...)
-	return nil
+
+	if err != nil {
+		err = fmt.Errorf("%w: %w; Channels: %s", stream.ErrSubscriptionFailure, err, strings.Join(cNames, ", "))
+		b.Websocket.DataHandler <- err
+	}
+
+	return err
 }
 
 // Unsubscribe unsubscribes from a set of channels
-func (b *Binance) Unsubscribe(channelsToUnsubscribe []subscription.Subscription) error {
-	payload := WsPayload{
-		Method: "UNSUBSCRIBE",
-	}
-	for i := range channelsToUnsubscribe {
-		payload.Params = append(payload.Params, channelsToUnsubscribe[i].Channel)
-		if i%50 == 0 && i != 0 {
-			err := b.Websocket.Conn.SendJSONMessage(payload)
-			if err != nil {
-				return err
-			}
-			payload.Params = []string{}
-		}
+func (b *Binance) Unsubscribe(channels []subscription.Subscription) error {
+	return b.ParallelChanOp(channels, b.unsubscribeFromChan, 50)
+}
+
+// unsubscribeFromChan sends a websocket message to stop receiving data from a channel
+func (b *Binance) unsubscribeFromChan(chans []subscription.Subscription) error {
+	id := b.Websocket.Conn.GenerateMessageID(false)
+
+	cNames := make([]string, len(chans))
+	for i := range chans {
+		cNames[i] = chans[i].Channel
 	}
-	if len(payload.Params) > 0 {
-		err := b.Websocket.Conn.SendJSONMessage(payload)
-		if err != nil {
-			return err
+
+	req := WsPayload{
+		Method: wsUnsubscribeMethod,
+		Params: cNames,
+		ID:     id,
+	}
+
+	respRaw, err := b.Websocket.Conn.SendMessageReturnResponse(id, req)
+	if err == nil {
+		if _, d, _, rErr := jsonparser.Get(respRaw, "result"); rErr != nil {
+			err = rErr
+		} else if d != jsonparser.Null { // null is the only expected and acceptable response
+			err = errUnknownError
 		}
 	}
-	b.Websocket.RemoveSubscriptions(channelsToUnsubscribe...)
+
+	if err != nil {
+		err = fmt.Errorf("%w: %w; Channels: %s", stream.ErrUnsubscribeFailure, err, strings.Join(cNames, ", "))
+		b.Websocket.DataHandler <- err
+	}
+
+	b.Websocket.RemoveSubscriptions(chans...)
+
 	return nil
 }
 
+// listSubscriptoins returns the current list of subscribed channels
+func (b *Binance) listSubscriptions() ([]string, error) {
```

Comment:
This isn't being used in the current work flow or being tested? Any chance of integrating this and adding it to the sub/unsub?

Reply:
See commit message 0c584dd9

Reply:
Note @thrasher-  @gloriousCode 👆 This is one of my best arguments for not squashing.
If I want to implement something, and then roll it back but show how it was done, I do it like this, and that commit in the history becomes important, even though it was reverted inside the PR.
Don't have a solution other than maybe labelling some PRs to be FF merged instead of squash merged

Reply:
@gbjk I think in future it's best to have those caveats included as a warning or note on the actual function comment, instead of diving into git history. 

If what you say is true and the fact that its not included in any other function, it's not exported and its not tested. Just remove it completely, it's dead code. Unless you want to use it as a library function export it and test it and add commentary.

Reply:
Okay; I'll remove it.

Regarding comments versus commits, I think both are warranted. The comment should explain why, but the commit would give a viable implementation should things change. A better approach is probably to punt it into a mothball/ branch though, and I'll do that now instead.